engine/execution/provider/engine.go

```diff
@@ -24,150 +28,165 @@ import (
+	chdpRequestProcessInterval time.Duration
 }
 
 func New(
 	logger zerolog.Logger,
 	tracer module.Tracer,
 	net network.Network,
 	state protocol.State,
-	me module.Local,
 	execState state.ReadOnlyExecutionState,
 	metrics module.ExecutionMetrics,
 	checkAuthorizedAtBlock func(blockID flow.Identifier) (bool, error),
-	chdpQueryTimeout uint,
-	chdpDeliveryTimeout uint,
+	chunkDataPackRequestQueue mempool.ChunkDataPackRequestQueue,
+	chdpQueryTimeout time.Duration,
+	chdpDeliveryTimeout time.Duration,
+	chdpRequestProcessInterval time.Duration,
+	chdpRequestWorkers uint,
 ) (*Engine, error) {
 
 	log := logger.With().Str("engine", "receipts").Logger()
 
-	eng := Engine{
-		unit:                   engine.NewUnit(),
-		log:                    log,
-		tracer:                 tracer,
-		state:                  state,
-		me:                     me,
-		execState:              execState,
-		metrics:                metrics,
-		checkAuthorizedAtBlock: checkAuthorizedAtBlock,
-		chdpQueryTimeout:       time.Duration(chdpQueryTimeout) * time.Second,
-		chdpDeliveryTimeout:    time.Duration(chdpDeliveryTimeout) * time.Second,
+	engine := Engine{
+		log:                        log,
+		tracer:                     tracer,
+		state:                      state,
+		execState:                  execState,
+		metrics:                    metrics,
+		checkAuthorizedAtBlock:     checkAuthorizedAtBlock,
+		chdpQueryTimeout:           chdpQueryTimeout,
+		chdpDeliveryTimeout:        chdpDeliveryTimeout,
+		chdpRequestProcessInterval: chdpRequestProcessInterval,
+		chdpRequestsQueue:          chunkDataPackRequestQueue,
 	}
 
 	var err error
 
-	eng.receiptCon, err = net.Register(channels.PushReceipts, &eng)
+	engine.receiptCon, err = net.Register(channels.PushReceipts, &engine)
 	if err != nil {
 		return nil, fmt.Errorf("could not register receipt provider engine: %w", err)
 	}
 
-	chunksConduit, err := net.Register(channels.ProvideChunks, &eng)
+	chunksConduit, err := net.Register(channels.ProvideChunks, &engine)
 	if err != nil {
 		return nil, fmt.Errorf("could not register chunk data pack provider engine: %w", err)
 	}
-	eng.chunksConduit = chunksConduit
-
-	return &eng, nil
-}
+	engine.chunksConduit = chunksConduit
 
-func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.ProcessLocal(event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	cm := component.NewComponentManagerBuilder()
+	for i := uint(0); i < chdpRequestWorkers; i++ {
+		cm.AddWorker(engine.processChunkDataPackRequestWorker)
+	}
 
-func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(channel, originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	engine.cm = cm.Build()
+	engine.Component = engine.cm
 
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), event)
-	})
```

Comment:
Upgraded `ProviderEngine` to `MessageProcessor` interface (from deprecated `Engine` interface), which no longer requires these methods.